Generic Host 3.0 update (#13072)
aspnet/AspNetCore.Docs: aspnetcore/fundamentals/host/generic-host.md

## Set up a host

The host is typically configured, built, and run by code in the `Program` class. The `Main` method:

* Calls a `CreateHostBuilder` method to create and configure a builder object.
* Calls `Build` and `Run` methods on the builder object.

Here's *Program.cs* code for a non-HTTP workload, with a single `IHostedService` implementation added to the DI container. 

```csharp
public class Program
{
    public static void Main(string[] args)
    {
        CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureServices((hostContext, services) =>
            {
               services.AddHostedService<Worker>();
            });
}
```

For an HTTP workload, the `Main` method is the same but `CreateHostBuilder` calls `ConfigureWebHostDefaults`:

```csharp
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.UseStartup<Startup>();
        });
```

> If the app uses Entity Framework Core, don't change the name of the `CreateHostBuilder` method. The [Entity Framework Core tools](/ef/core/miscellaneous/cli/) expect to find a `CreateHostBuilder` method that configures the host without running the app. For more information, see [Design-time DbContext Creation](/ef/core/miscellaneous/cli/dbcontext-creation).

Reviewer: ```suggestion
If the app uses Entity Framework Core, don't change the name or signature of the `CreateHostBuilder` method. The [Entity Framework Core tools](/ef/core/miscellaneous/cli/) expect to find a `CreateHostBuilder` method that configures the host without running the app. For more information, see [Design-time DbContext Creation](/ef/core/miscellaneous/cli/dbcontext-creation).
```

## Default builder settings 

The <xref:Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder*> method:

* Sets the content root to the path returned by <xref:System.IO.Directory.GetCurrentDirectory*>.

> * Loads host configuration from environment variables prefixed with "DOTNET_".

Reviewer: And command line arguments.

* Loads app configuration from:
  * *appsettings.json*.
  * *appsettings.{Environment}.json*.
  * [Secret Manager](xref:security/app-secrets) when the app runs in the `Development` environment.
  * Environment variables.
  * Command-line arguments.
* Adds the following [logging](xref:fundamentals/logging/index) providers:
  * Console
  * Debug
  * EventSource
  * EventLog (only when running on Windows)

> * Enables [scope validation](xref:fundamentals/dependency-injection#scope-validation) when the environment is Development.

Reviewer: and dependency validation
https://github.com/aspnet/Extensions/blob/689e0d7644d5c92c1ce2f33cfc02955fa254d7f3/src/DependencyInjection/DI/src/ServiceProviderOptions.cs#L21-L25

[!code-csharp[](generic-host/samples-snapshot/3.x/LifetimeEventsHostedService.cs?name=snippet_LifetimeEvents)]

## IHostLifetime

The <xref:Microsoft.Extensions.Hosting.IHostLifetime> implementation controls when the host starts and when it stops. The last implementation registered is used.

<xref:Microsoft.Extensions.Hosting.Internal.ConsoleLifetime> is the default `IHostLifetime` implementation. `ConsoleLifetime`:

* listens for Ctrl+C/SIGINT or SIGTERM and calls <xref:Microsoft.Extensions.Hosting.IApplicationLifetime.StopApplication*> to start the shutdown process.
* Unblocks extensions such as [RunAsync](#runasync) and [WaitForShutdownAsync](#waitforshutdownasync).

## IHostEnvironment

Inject the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> service into a class to get information about the following:

* [ApplicationName](#applicationname)
* [EnvironmentName](#environmentname)
* [ContentRootPath](#contentrootpath)

Web apps implement the `IWebHostEnvironment` interface, which inherits `IHostEnvironment` and adds:

* [WebRootPath](#webroot)

## Host configuration

> Host configuration is used for the following:

Reviewer: It's available from HostBuilderContext.Configuration inside ConfigureAppConfiguration. After ConfigureAppConfiguration HostBuilderContext.Configuration is replaced with the app config.
https://github.com/aspnet/Extensions/blob/36a9bf9d3fdea32b548d7001b27e6b19ae78b094/src/Hosting/Hosting/src/HostBuilder.cs#L169-L190

* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.

> * The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.

Reviewer: IServerAddresses is pulled from app configuration for generic host.

Host configuration automatically flows to [app configuration](#app-configuration).

To add host configuration, call <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureHostConfiguration*> on `IHostBuilder`. `ConfigureHostConfiguration` can be called multiple times with additive results. The host uses whichever option sets a value last on a given key.

> The environment variable provider with prefix `DOTNET_` is included by default. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.

Reviewer: ```suggestion
The environment variable provider with prefix `DOTNET_` and command line args are included by CreateDefaultBuilder. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.
```

```csharp
webBuilder.UseSetting(WebHostDefaults.PreventHostingStartupKey, "true");
```

### StartupAssembly

The assembly to search for the `Startup` class.

**Key**: startupAssembly
**Type**: *string*  
**Default**: The app's assembly  
**Environment variable**: `<PREFIX_>STARTUPASSEMBLY`

To set this value, use the environment variable or call `UseStartup`. `UseStartup` can take an assembly name (`string`) or a type (`TStartup`). If multiple `UseStartup` methods are called, the last one takes precedence.

```csharp
webBuilder.UseStartup("StartupAssemblyName");
```

```csharp
webBuilder.UseStartup<Startup>();
```

### URLs

A semicolon-delimited list of IP addresses or host addresses with ports and protocols that the server should listen on for requests. For example, `http://localhost:123`. Use "\*" to indicate that the server should listen for requests on any IP address or hostname using the specified port and protocol (for example, `http://*:5000`). The protocol (`http://` or `https://`) must be included with each URL. Supported formats vary among servers.

**Key**: urls  
**Type**: *string*  
**Default**: `http://localhost:5000` and `https://localhost:5001`

> **Environment variable**: `ASPNETCORE_URLS`

Reviewer: ```suggestion
**Environment variable**: `<PREFIX_>URLS`
```

To set this value, use the environment variable or call `UseUrls`:

```csharp
webBuilder.UseUrls("http://*:5000;http://localhost:5001;https://hostname:5002");
```

Kestrel has its own endpoint configuration API. For more information, see <xref:fundamentals/servers/kestrel#endpoint-configuration>.

### WebRoot

The relative path to the app's static assets.

**Key**: webroot  
**Type**: *string*  
**Default**: *(Content Root)/wwwroot*, if the path exists. If the path doesn't exist, a no-op file provider is used.  

> **Environment variable**: `ASPNETCORE_WEBROOT`

Reviewer: ```suggestion
**Environment variable**: `<PREFIX_>WEBROOT`
```

### RunAsync

<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync*> runs the app and returns a <xref:System.Threading.Tasks.Task> that completes when the cancellation token or shutdown is triggered.

### RunConsoleAsync

<xref:Microsoft.Extensions.Hosting.HostingHostBuilderExtensions.RunConsoleAsync*> enables console support, builds and starts the host, and waits for Ctrl+C/SIGINT or SIGTERM to shut down.

### Start

<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Start*> starts the host synchronously.

### StartAsync

<xref:Microsoft.Extensions.Hosting.IHost.StartAsync*> starts the host and returns a <xref:System.Threading.Tasks.Task> that completes when the cancellation token or shutdown is triggered. 

<xref:Microsoft.Extensions.Hosting.IHostLifetime.WaitForStartAsync*> is called at the start of `StartAsync`, which waits until it's complete before continuing. This can be used to delay startup until signaled by an external event.

### StopAsync

<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.StopAsync*> attempts to stop the host within the provided timeout.

### WaitForShutdown

> <xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdown*> blocks the calling thread until shutdown is triggered via Ctrl+C/SIGINT or SIGTERM.

Reviewer: ```suggestion
<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdown*> blocks the calling thread until shutdown is triggered by the IHostLifetime, such as via Ctrl+C/SIGINT or SIGTERM.
```

### WaitForShutdownAsync

<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdownAsync*> returns a <xref:System.Threading.Tasks.Task> that completes when shutdown is triggered via the given token and calls <xref:Microsoft.Extensions.Hosting.IHost.StopAsync*>.

### External control

> External control of the host can be achieved using methods that can be called externally:

Reviewer: ```suggestion
Direct control of the host lifetime can be achieved using methods that can be called externally:
```

<xref:Microsoft.Extensions.Hosting.Internal.ConsoleLifetime> is the default `IHostLifetime` implementation. `ConsoleLifetime`:

* listens for Ctrl+C/SIGINT or SIGTERM and calls <xref:Microsoft.Extensions.Hosting.IApplicationLifetime.StopApplication*> to start the shutdown process.
* Unblocks extensions such as [RunAsync](#runasync) and [WaitForShutdownAsync](#waitforshutdownasync).

## IHostEnvironment

Inject the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> service into a class to get information about the following:

* [ApplicationName](#applicationname)
* [EnvironmentName](#environmentname)
* [ContentRootPath](#contentrootpath)

Web apps implement the `IWebHostEnvironment` interface, which inherits `IHostEnvironment` and adds:

* [WebRootPath](#webroot)

## Host configuration

Host configuration is used for the following:

* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.
* The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.

> Host configuration automatically flows to app configuration ([ConfigureAppConfiguration](#configureappconfiguration).

Reviewer 2: Host configuration automatically flows to app configuration ([ConfigureAppConfiguration](#configureappconfiguration)).

Reviewer 2: Or this next version might be better

Reviewer 2: Host configuration automatically flows to [app configuration](#configureappconfiguration).

To add host configuration, call <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureHostConfiguration*> on `IHostBuilder`. `ConfigureHostConfiguration` can be called multiple times with additive results. The host uses whichever option sets a value last on a given key.

The environment variable provider with prefix `DOTNET_` is included by default. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.

The following example creates host configuration:

[!code-csharp[](generic-host/samples-snapshot/3.x/Program.cs?name=snippet_HostConfig)]

> > [!TIP]

Reviewer 2: Too much bling for me. This also seems TMI for an advanced topic. Perhaps something like my next proposal.

> > [!TIP]

Reviewer 2: Development environment variables can be set in *.json* files.  For more information, see <xref:fundamentals/environments>.

Author: Environments doc is missing some of what's in the current text; I'll condense the text and drop the Tip box.

Reviewer 2: Can you open an issue on the environment doc? That level of detail doesn't belong in this doc.

Author: Agreed on both counts.

> During development when using [Visual Studio](https://visualstudio.microsoft.com) or running an app with `dotnet run`, environment variables may be set in the *Properties/launchSettings.json* file. In [Visual Studio Code](https://code.visualstudio.com/), environment variables may be set in the *.vscode/launch.json* file. For more information, see <xref:fundamentals/environments>.

## App configuration

App configuration is created by calling <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureAppConfiguration*> on `IHostBuilder`. `ConfigureAppConfiguration` can be called multiple times with additive results. The app uses whichever option sets a value last on a given key. 

The configuration created by `ConfigureAppConfiguration` is available at [HostBuilderContext.Configuration](xref:Microsoft.Extensions.Hosting.HostBuilderContext.Configuration*) for subsequent operations and as a service from DI.

App configuration automatically receives host configuration provided by `ConfigureHostConfiguration`.

The following example creates app configuration:

[!code-csharp[](generic-host/samples-snapshot/3.x/Program.cs?name=snippet_AppConfig)]

> > [!TIP]

Reviewer 2: Not a tip bling fan. The content is fine.

> To move settings files to the output directory for publishing, specify the settings files as [MSBuild project items](/visualstudio/msbuild/common-msbuild-project-items) in the project file. The sample app moves its JSON app settings files and *hostsettings.json* with the following `<Content>` item:
> ```xml
> <ItemGroup>
>   <Content Include="**\*.json" Exclude="bin\**\*;obj\**\*" 
>     CopyToOutputDirectory="PreserveNewest" />
> </ItemGroup>
> ```

## Settings for all app types

> This section lists host settings that apply to non-HTTP as well as HTTP workloads. By default, environment variables used to configure these settings can have a `DOTNET_` or `ASPNETCORE_` prefix.

Reviewer 2: This section lists host settings that apply to non-HTTP and HTTP workloads. By default, environment variables used to configure these settings can have a `DOTNET_` or `ASPNETCORE_` prefix.

## What's a host?

A *host* is an object that encapsulates an app's resources, such as:

* Dependency injection (DI)
* Logging
* Configuration
* `IHostedService` implementations

When a host starts, it calls `IHostedService.StartAsync` on each implementation of <xref:Microsoft.Extensions.Hosting.IHostedService> that it finds in the DI container. In a web app, one of the `IHostedService` implementations is a web service that starts an [HTTP server implementation](xref:fundamentals/index#servers).

The main reason for including all of the app's interdependent resources in one object is lifetime management: control over app startup and graceful shutdown.

> ## Web Host is deprecated

Reviewer: @glennc for messaging?
